p2p/stream/protocols/sync/client.go

```diff
@@ -0,0 +1,344 @@
+func (p *Protocol) GetBlockHashes(ctx context.Context, bns []uint64, opts ...Option) ([]common.Hash, sttypes.StreamID, error) {
+	if len(bns) > GetBlockHashesAmountCap {
+		return nil, "", fmt.Errorf("number of requested numbers exceed limit")
+	}
+
+	req := newGetBlockHashesRequest(bns)
+	resp, stid, err := p.rm.DoRequest(ctx, req, opts...)
+	if err != nil {
+		return nil, stid, err
+	}
+	hashes, err := req.getHashesFromResponse(resp)
+	if err != nil {
+		return nil, stid, err
+	}
+	return hashes, stid, nil
+}
+
+// GetBlocksByHashes do getBlocksByHashesRequest through sync stream protocol.
+func (p *Protocol) GetBlocksByHashes(ctx context.Context, hs []common.Hash, opts ...Option) ([]*types.Block, sttypes.StreamID, error) {
+	if len(hs) > GetBlocksByHashesAmountCap {
+		return nil, "", fmt.Errorf("number of requested hashes exceed limit")
+	}
+	req := newGetBlocksByHashesRequest(hs)
+	resp, stid, err := p.rm.DoRequest(ctx, req, opts...)
+	if err != nil {
+		return nil, stid, err
+	}
+	blocks, err := req.getBlocksFromResponse(resp)
+	if err != nil {
+		return nil, stid, err
+	}
+	return blocks, stid, nil
+}
+
+// getBlocksByNumberRequest is the request for get block by numbers which implements
+// sttypes.Request interface
+type getBlocksByNumberRequest struct {
+	bns   []uint64
+	pbReq *syncpb.Request
+}
+
+func newGetBlocksByNumberRequest(bns []uint64) *getBlocksByNumberRequest {
+	pbReq := syncpb.MakeGetBlocksByNumRequest(bns)
+	return &getBlocksByNumberRequest{
+		bns:   bns,
+		pbReq: pbReq,
+	}
+}
+
+func (req *getBlocksByNumberRequest) ReqID() uint64 {
+	return req.pbReq.GetReqId()
+}
+
+func (req *getBlocksByNumberRequest) SetReqID(val uint64) {
+	req.pbReq.ReqId = val
+}
+
+func (req *getBlocksByNumberRequest) String() string {
+	ss := make([]string, 0, len(req.bns))
+	for _, bn := range req.bns {
+		ss = append(ss, strconv.Itoa(int(bn)))
+	}
+	bnsStr := strings.Join(ss, ",")
+	return fmt.Sprintf("REQUEST [GetBlockByNumber: %s]", bnsStr)
+}
+
+func (req *getBlocksByNumberRequest) IsSupportedByProto(target sttypes.ProtoSpec) bool {
+	return target.Version.GreaterThanOrEqual(MinVersion)
+}
+
+func (req *getBlocksByNumberRequest) Encode() ([]byte, error) {
+	msg := syncpb.MakeMessageFromRequest(req.pbReq)
+	return protobuf.Marshal(msg)
+}
+
+func (req *getBlocksByNumberRequest) getBlocksFromResponse(resp sttypes.Response) ([]*types.Block, error) {
+	sResp, ok := resp.(*syncResponse)
+	if !ok || sResp == nil {
+		return nil, errors.New("not sync response")
+	}
+	blockBytes, sigs, err := req.parseBlockBytesAndSigs(sResp)
+	if err != nil {
+		return nil, err
+	}
+	blocks := make([]*types.Block, 0, len(blockBytes))
+	for i, bb := range blockBytes {
+		var block *types.Block
+		if err := rlp.DecodeBytes(bb, &block); err != nil {
+			return nil, errors.Wrap(err, "[GetBlocksByNumResponse]")
+		}
+		if block != nil {
+			block.SetCurrentCommitSig(sigs[i])
+		}
+		blocks = append(blocks, block)
```

Comment:
if block is nil, keep adding it?

Reply:
Yes. nil is a placeholder equivalent to say "I don't have the block message of the given hash/number".